cmd/launcher/main.go

```diff
@@ -270,186 +266,102 @@ func main() {
+		"version", versionInfo.Version,
+		"build", versionInfo.Revision,
 	)
-	if err != nil {
-		logger.Fatal(errors.Wrap(err, "launching osquery instance"))
-	}
-
-	// The runner allows querying the osqueryd instance from the extension.
-	// Used by the Enroll method below to get initial enrollment details.
-	ext.SetQuerier(runner)
-
-	ctx := context.Background()
-	_, invalid, err := ext.Enroll(ctx)
-	if err != nil {
-		logger.Fatal("err", errors.Wrap(err, "enrolling host"))
-	}
-	if invalid {
-		logger.Fatal(errors.Wrap(err, "invalid enroll secret"))
-	}
-	ext.Start()
-	defer ext.Shutdown()
 
 	// If the control server has been opted-in to, run it
 	if opts.control {
-		level.Debug(logger).Log("msg", "creating control client")
-
-		controlOpts := []control.Option{
-			control.WithLogger(logger),
-			control.WithGetShellsInterval(opts.getShellsInterval),
-		}
-		if opts.insecureTLS {
-			controlOpts = append(controlOpts, control.WithInsecureSkipVerify())
-		}
-		if opts.disableControlTLS {
-			controlOpts = append(controlOpts, control.WithDisableTLS())
-		}
-		controlClient, err := control.NewControlClient(db, opts.controlServerURL, controlOpts...)
+		control, err := createControl(ctx, db, logger, opts)
 		if err != nil {
-			logger.Fatal(errors.Wrap(err, "starting control client"))
+			logger.Fatal(errors.Wrap(err, "creating control actor"))
 		}
-
-		controlClient.Start(ctx)
-		defer controlClient.Stop()
+		runGroup.Add(control.Execute, control.Interrupt)
 	}
 
 	// If the autoupdater is enabled, enable it for both osquery and launcher
 	if opts.autoupdate {
-		enabler := &updateEnabler{
+		config := &updaterConfig{
 			Logger:             logger,
 			RootDirectory:      rootDirectory,
 			AutoupdateInterval: opts.autoupdateInterval,
+			UpdateChannel:      opts.updateChannel,
 			NotaryURL:          opts.notaryServerURL,
 			MirrorURL:          opts.mirrorServerURL,
-			HttpClient:         httpClient,
+			HTTPClient:         httpClient,
 		}
 
-		stopOsquery, err := enabler.EnableBinary(
-			opts.osquerydPath,
-			autoupdate.WithFinalizer(runner.Restart),
-			autoupdate.WithUpdateChannel(opts.updateChannel),
-		)
+		// create an updater for osquery
+		osqueryUpdater, err := createUpdater(ctx, opts.osquerydPath, extension.Restart, logger, config)
 		if err != nil {
 			logger.Fatal(err)
 		}
-		defer stopOsquery()
+		runGroup.Add(osqueryUpdater.Execute, osqueryUpdater.Interrupt)
 
 		launcherPath, err := os.Executable()
 		if err != nil {
 			logger.Fatal(err)
 		}
-		stopLauncher, err := enabler.EnableBinary(
+		launcherUpdater, err := createUpdater(
+			ctx,
 			launcherPath,
-			autoupdate.WithFinalizer(launcherFinalizer(logger, runner.Shutdown)),
-			autoupdate.WithUpdateChannel(opts.updateChannel),
+			launcherFinalizer(logger, extension.Shutdown),
+			logger,
+			config,
 		)
 		if err != nil {
 			logger.Fatal(err)
 		}
-		defer stopLauncher()
+		runGroup.Add(launcherUpdater.Execute, launcherUpdater.Interrupt)
 	}
 
+	// start the rungroup
```

Comment:
starting the group should block. 
Add the Signal handler as an actor as well instead of launching this goroutine. 

Something like this:

```
			var g run.Group
			{
				// this actor handles an os interrupt signal and terminates the server.
				sig := make(chan os.Signal, 1)
				g.Add(func() error {
					signal.Notify(sig, os.Interrupt)
					select {
					case <-sig:
						level.Info(logger).Log("msg", "beginnning shutdown")
						return nil
					}
				}, func(err error) {
					level.Info(logger).Log("msg", "interrupted", "err", err)
					close(sig)
				})
			}
```